packages/generator/src/templateHelpers.ts

```diff
@@ -243,68 +258,96 @@ export const getValue = function(name: any): string {
   return null;
 };
 
-const getProperties = function(
-  propertyShapes: model.domain.PropertyShape[],
-  filterEntriesBy: Function
+type propertyFilter = (propertyName: string) => boolean;
+
+/**
+ * Get properties of the DTO (inherited and linked) after applying the given filter criteria
+ *
+ * @param dtoTypeModel AMF model of the dto
+ * @param propertyFilter function to filter properties based on certain criteria
+ */
+const getFilteredProperties = function(
+  dtoTypeModel: model.domain.NodeShape | null | undefined,
+  propertyFilter: propertyFilter
 ): model.domain.PropertyShape[] {
-  return !propertyShapes
-    ? []
-    : propertyShapes.filter(entry => filterEntriesBy(entry));
+  const properties: model.domain.PropertyShape[] = [];
+  const existingProps: Set<string> = new Set();
+
+  while (dtoTypeModel != null) {
+    if (dtoTypeModel.properties != null && dtoTypeModel.properties.length > 0) {
+      dtoTypeModel.properties.forEach(prop => {
+        if (prop != null) {
+          const propName = getValue(prop.name);
+          //ignore duplicate props
+          if (
+            propName != null &&
+            !existingProps.has(propName) &&
+            propertyFilter(propName) === true
+          ) {
+            existingProps.add(propName);
+            properties.push(prop);
+          }
+        }
+      });
+      //Check if there are any inherited properties
+      if (dtoTypeModel.inherits != null && dtoTypeModel.inherits.length > 0) {
```

Comment:
I remember you were mentioning about cyclic references and also resolving it 

I am still not clear on how this while loop will complete for cyclic references.

For Example given 2 types as below
Type1 {
   variable: Type2;
}

Type2 {
   variable: Type1
}

If block might always set the model to a valid NodeShape..

Will stop by to discuss